services/motion/builtin/move_request.go

```diff
@@ -85,12 +90,88 @@ func (mr *moveRequest) deviatedFromPlan(ctx context.Context, waypoints [][]refer
 	if err != nil {
 		return false, err
 	}
-	mr.planRequest.Logger.Debug("deviation from plan: %v", errorState.Point())
+	mr.planRequest.Logger.Debugf("deviation from plan: %v", errorState.Point())
 	return errorState.Point().Norm() > mr.config.PlanDeviationMM, nil
 }
 
 func (mr *moveRequest) obstaclesIntersectPlan(ctx context.Context, waypoints [][]referenceframe.Input, waypointIndex int) (bool, error) {
-	// TODO(RSDK-4507): implement this function
+	// convert waypoints to type of motionplan.Plan
+	var plan []map[string][]referenceframe.Input
+	// We only care to check against waypoints we have not reached yet.
+	for _, inputs := range waypoints[waypointIndex:] {
+		input := make(map[string][]referenceframe.Input)
+		input[mr.kinematicBase.Name().Name] = inputs
+		plan = append(plan, input)
+	}
+
+	// iterate through vision services
+	for _, srvc := range mr.visionSvcs {
+		// iterate through all cameras on the robot
+		// any camera may be used by any vision service
+		for _, camName := range mr.cameraNames {
+			// get detections from vision service
+			detections, err := srvc.GetObjectPointClouds(ctx, camName.Name, nil)
```

Comment:
Is there really no better way to tell what vision service has what camera than trying *every* camera against *every* vision service?

Reply:
Every vision service has access to every camera.

Reply:
I'm not sure that will always be true. If viz service A is on a remote, and camera B is on the local robot, then A would not be able to see B.

Reply:
> If viz service A is on a remote, and camera B is on the local robot, then A would not be able to see B.

Correct imo. When you're iterating through "all cameras" on the robot, are you wanting to see cameras of main robots that point to this robot as a remote? I don't think that's possible with our current architecture, but am curious if that's a feature you need.

Reply:
good catch, I wasn't even thinking about how remotes might play a part in this.
In this case we will need to know, when the motion service is reconfigured, which if any of the services and components is remote.
to this end, we can accomplish this with
https://github.com/viamrobotics/rdk/blob/main/resource/name.go#L92-L95